### PR TITLE
Fix starttls emitting double connected events

### PR DESCRIPTION
**Pull request by joshperry**
Fix double event issue in #19232 

When using the starttls filter, it breaks the expectation that connections should only raise the connected event once. This expectation is relied on for proper operation of core envoy functions like the connection pooler.

Risk Level: Low

Testing: Added regression test.

### CODE REVIEW

**Comment by phlax** [comment]
hi @joshperry youll need to fix DCO on last commit

**Comment by phlax** [comment]
seems like CI errors are genuine

File: source/extensions/transport_sockets/starttls/starttls_socket.h [hunk 1]
```diff
@@ -50,6 +51,37 @@ class StartTlsSocket : public Network::TransportSocket, Logger::Loggable<Logger:
   bool startSecureTransport() override;
 
 private:
+  // This is a proxy for wrapping the transport callback object passed from the consumer
+  // Its primary purpose is to filter Connected events to ensure they only happen once per open
```
Comment (cpakulski) [inline source/extensions/transport_sockets/starttls/starttls_socket.h:55]: Please add dots at the end of the lines.

File: source/extensions/transport_sockets/starttls/starttls_socket.h [hunk 2]
```diff
@@ -50,6 +51,37 @@ class StartTlsSocket : public Network::TransportSocket, Logger::Loggable<Logger:
...
+        isopen_ = false;
+      }
+
+      parent_->raiseEvent(event);
+    }
+    void flushWriteBuffer() override { parent_->flushWriteBuffer(); }
+
+  private:
+    Network::TransportSocketCallbacks* parent_;
+    bool isopen_{false};
```
Comment (cpakulski) [inline source/extensions/transport_sockets/starttls/starttls_socket.h:82]: Rename to is_open_

**Review by cpakulski — CHANGES_REQUESTED** [review]
Overall looks good. Thanks for finding it and fixing.

You need to add unit tests for `CallbackProxy` to verify that each call results in a call to `parent_`'s method. In case of `raiseEvent` you should verify that parent is called only once when `raiseEvent(Network::ConnectionEvent::Connected)` is called twice. Those unit tests will bring coverage to higher percentage.

**Comment by joshperry** [comment]
Appreciate the input everyone! Been crunching this week, but I'll be able to finish this up after the holidays.

Cheers!

**Comment by lizan** [comment]
/wait
Ping me when this is ready.

**Comment by joshperry** [comment]
Have tests ready for this, not sure how to reopen.

**Pull request closed without merging**